### PR TITLE
PEBS monitor frequency

### PR DESCRIPTION
**Pull request by bratpiorka**
### Types of changes

- [ ] Bugfix (non-breaking change which fixes issue linked in Description above)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)
- [ ] Other

### Checklist

- [ ] Code compiles without errors
- [ ] All tests pass locally with my changes (see TESTING section in CONTRIBUTING file)
- [ ] Created tests which will fail without the change (if possible)
- [ ] Extended the README/documentation (if necessary)
- [ ] All newly added files have proprietary license (if necessary)
- [ ] All newly added files are referenced in MANIFEST files (if necessary)

## Further comments

---
This change is [Reviewable](https://reviewable.io/reviews/memkind/memkind/786)

### CODE REVIEW

**Review by PatKamin — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 4 unresolved discussions (waiting on @bratpiorka and @maciekpac)

___
*[test/memkind_memtier_data_movement_test.cpp, line 710 at r1](https://reviewable.io/reviews/memkind/memkind/786#-MzUA-8zBn3Exr2DW-hr:-MzUA-8zBn3Exr2DW-hs:b-ycy7vb) ([raw file](https://github.com/memkind/memkind/blob/4a1a0b1e0aa43cf45ef2967b9d9d259b157c1c82/test/memkind_memtier_data_movement_test.cpp#L710)):*
> ```cpp
> 
>     // Wait a little to allow PEBS thread do it's work
>     sleep(5);
> ```

Please, verify by running this test like 1000 times if this value can be lowered.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 4 unresolved discussions (waiting on @kilobyte, @maciekpac, and @PatKamin)

___
*[test/memkind_memtier_data_movement_test.cpp, line 709 at r1](https://reviewable.io/reviews/memkind/memkind/786#-MzU4Vgz0TybImvlwg8U:-MzUnmeM6zT_snU_9Rv6:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/4a1a0b1e0aa43cf45ef2967b9d9d259b157c1c82/test/memkind_memtier_data_movement_test.cpp#L709)):*
<details><summary><i>Previously, kilobyte (Adam Borowski) wrote…</i></summary><blockquote>

grammar nit: "its"
</blockquote></details>

Done.
___
*[test/memkind_memtier_data_movement_test.cpp, line 710 at r1](https://reviewable.io/reviews/memkind/memkind/786#-MzUA-8zBn3Exr2DW-hr:-MzUnnCx0RnmhKWSQfYH:be178le) ([raw file](https://github.com/memkind/memkind/blob/4a1a0b1e0aa43cf45ef2967b9d9d259b157c1c82/test/memkind_memtier_data_movement_test.cpp#L710)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

Please, verify by running this test like 1000 times if this value can be lowered.
</blockquote></details>

2 sec is enough, checked > 100 test iterations
___
*[test/memkind_memtier_data_movement_test.cpp, line 728 at r1](https://reviewable.io/reviews/memkind/memkind/786#-MzU4brJAOHHFO8nq2EC:-MzUo2tZ0KPiYCC1kx_E:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/4a1a0b1e0aa43cf45ef2967b9d9d259b157c1c82/test/memkind_memtier_data_movement_test.cpp#L728)):*
<details><summary><i>Previously, kilobyte (Adam Borowski) wrote…</i></summary><blockquote>

its
</blockquote></details>

Done.
___
*[test/memkind_memtier_data_movement_test.cpp, line 778 at r1](https://reviewable.io/reviews/memkind/memkind/786#-MzU4dVPC00AyRGXbX_W:-MzUo3EQEW3SCO6DowlM:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/4a1a0b1e0aa43cf45ef2967b9d9d259b157c1c82/test/memkind_memtier_data_movement_test.cpp#L778)):*
<details><summary><i>Previously, kilobyte (Adam Borowski) wrote…</i></summary><blockquote>

its
</blockquote></details>

Done.

**Review by PatKamin — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 3 unresolved discussions (waiting on @kilobyte and @maciekpac)

File: test/memkind_memtier_data_movement_test.cpp [hunk 1]
```diff
@@ -706,6 +706,8 @@ TEST_F(PEBSTest, Basic)
     actual_size = ranking_get_total_size(mtt_allocator.internals.dramRanking);
     ASSERT_EQ(actual_size, size);
 
+    // Wait a little to allow PEBS thread do its work
+    sleep(2);
```
Comment (maciekpac) [inline test/memkind_memtier_data_movement_test.cpp:710]: Please, don't use ```sleep``` for thread synchronization, even if tests, unless really necessary.

There is a function that was written specifically for this purpose:
```C
    mtt_allocator_await_flush(&mtt_allocator);
```
Also, making the sleep non-interruptible significatly increases the execution time of any code that depends on flushing. Instead of calling ```sleep``` in pebs.c, I think we should call 
```C
        pthread_cond_timedwait(&mtt_allocator->bgThread.bg_thread_cond,
                          &mtt_allocator->bgThread.bg_thread_cond_mutex, wait_time);
```
in ```mtt_run_bg_thread```. This would automagically be interrupted by any call to ```mtt_allocator_await_flush``` (or ```mtt_allocator_await_balance```).

File: test/memkind_memtier_data_movement_test.cpp [hunk 2]
```diff
@@ -723,6 +725,8 @@ TEST_F(PEBSTest, Basic)
         }
     }
 
+    // Wait a little to allow PEBS thread do its work
+    sleep(2);
```
Comment (maciekpac) [inline test/memkind_memtier_data_movement_test.cpp:729]: Ditto

File: test/memkind_memtier_data_movement_test.cpp [hunk 3]
```diff
@@ -771,6 +775,8 @@ TEST_F(PEBSTest, SoftLimitMovementLogic)
...
```
Comment (maciekpac) [inline test/memkind_memtier_data_movement_test.cpp:779]: Ditto

**Review by maciekpac — CHANGES_REQUESTED** [review]

File: test/memkind_memtier_data_movement_test.cpp (hunk 1 above)
Comment (maciekpac) [reply test/memkind_memtier_data_movement_test.cpp:710]: ```pthread_cond_clockwait``` would be more appropriate, in our case

**Review by maciekpac — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 9 unresolved discussions (waiting on @bratpiorka and @kilobyte)

___
*[src/mtt_allocator.c, line 14 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzVriGE2tIgzNVjGDGa:-MzVriGE2tIgzNVjGDGb:b-opw5kz) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L14)):*
> ```C
> #include <stdlib.h>
> 
> #define BG_THREAD_FREQUENCY 2.0
> ```

I think we would have to check performance with different frequencies, but if the sleep is interruptible, I think we might want to give some bigger values, like 0.3 or 0.1, but this is just a general comment - it's not necessary to change this value right now.
___
*[src/mtt_allocator.c, line 41 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzVsJLI8wlwMhcyx9Ij:-MzVsJLI8wlwMhcyx9Ik:b84nr1b) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L41)):*
> ```C
> static void unregister_mtt_allocator(MTTAllocator *mtt_allocator);
> 
> static void timespec_add(struct timespec *modified,
> ```

Nit: maybe we could keep these functions in a separate file ? They are pretty generic and unrelated to mtt_allocator itself. Also, this would let us easily write unit tests, in case we need to debug some weird corner cases with double precision in the future,
___
*[src/mtt_allocator.c, line 60 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzVssJX5AgrBRuL3OUy:-MzVssJX5AgrBRuL3OUz:b-qshlr3) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L60)):*
> ```C
>     struct timespec start_time, end_time;
>     // calculate time period to wait
>     double bg_freq_hz = BG_THREAD_FREQUENCY;
> ```

nit: why do we declare a variable instead of directly using the macro ?
___
*[test/memkind_memtier_data_movement_test.cpp, line 710 at r2](https://reviewable.io/reviews/memkind/memkind/786#-MzQLedocopfwCNvpEbG-r2-710:-MzVvnqj8WHctWdsn-sH:bbcnb8s) ([raw file](https://github.com/memkind/memkind/blob/ac2d7dcb14017eba637f3d8d53d1fa9277d3b6f8/test/memkind_memtier_data_movement_test.cpp#L710)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

I'm not sure about calling await_flush here - in this case we have to wait here for collect PEBS data + update rankings based on that. If the rankings will be empty (pebs_monitor wouldn't be called for example), then await_flush wouldn't do the job.
</blockquote></details>

mtt_allocator_await_flush performs all scheduled tasks in the following order:

1. calls ```pebs_monitor```
2. adds pages to rankings
3. Updates hotness of all pages added to the ranking

So we should:

1. Await flush after malloc (done)
2. Perform some actions that will be visible for pebs (not sure if it's done here)
3. Await flush again

After this sequence, pages should already be moved according to gathered pebs data

**Review by kilobyte — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 7 unresolved discussions (waiting on @bratpiorka)

___
*[src/mtt_allocator.c, line 233 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzZMvZ37V3SWKmiKzGN:-MzZMJey1tqsae1ac_W2:bot7567) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L233)):*
> ```C
> }
> 
> static struct timespec timespec_diff(const struct timespec *time1,
> ```

Elsewhere, I found it much easier to convert the timespec value to int64_t instead of doing all these complex checks.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 6 of 6 files at r5, 1 of 1 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 5 unresolved discussions (waiting on @kilobyte and @maciekpac)

___
*[src/mtt_allocator.c, line 14 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzVriGE2tIgzNVjGDGa:-MzoTDaU16U1nhrajo7k:b32wtd) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L14)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

I think we would have to check performance with different frequencies, but if the sleep is interruptible, I think we might want to give some bigger values, like 0.3 or 0.1, but this is just a general comment - it's not necessary to change this value right now.
</blockquote></details>

ok
___
*[src/mtt_allocator.c, line 233 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzZMvZ37V3SWKmiKzGN:-MzpgszB8xF0gnW_Wnfs:b-yzqsar) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L233)):*
<details><summary><i>Previously, kilobyte (Adam Borowski) wrote…</i></summary><blockquote>

Elsewhere, I found it much easier to convert the timespec value to int64_t instead of doing all these complex checks.
</blockquote></details>

I don't see any unnecessary checks?
___
*[test/memkind_memtier_data_movement_test.cpp, line 710 at r2](https://reviewable.io/reviews/memkind/memkind/786#-MzQLedocopfwCNvpEbG-r2-710:-MzphAW1Boj-xXQjwtjP:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/ac2d7dcb14017eba637f3d8d53d1fa9277d3b6f8/test/memkind_memtier_data_movement_test.cpp#L710)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

mtt_allocator_await_flush performs all scheduled tasks in the following order:

1. calls ```pebs_monitor```
2. adds pages to rankings
3. Updates hotness of all pages added to the ranking

So we should:

1. Await flush after malloc (done)
2. Perform some actions that will be visible for pebs (not sure if it's done here)
3. Await flush again

After this sequence, pages should already be moved according to gathered pebs data
</blockquote></details>

Done.
___
*[test/memkind_memtier_data_movement_test.cpp, line 729 at r2](https://reviewable.io/reviews/memkind/memkind/786#-MzQLedocopfwCNvpEbG-r2-729:-MzphBa3-AUARa1AnNwr:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/ac2d7dcb14017eba637f3d8d53d1fa9277d3b6f8/test/memkind_memtier_data_movement_test.cpp#L729)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.
___
*[test/memkind_memtier_data_movement_test.cpp, line 779 at r2](https://reviewable.io/reviews/memkind/memkind/786#-MzQLedocopfwCNvpEbG-r2-779:-MzphN9QCK5Q8M39NHWg:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/ac2d7dcb14017eba637f3d8d53d1fa9277d3b6f8/test/memkind_memtier_data_movement_test.cpp#L779)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

Ditto
</blockquote></details>

Done.

**Review by maciekpac — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: 7 of 8 files reviewed, 10 unresolved discussions (waiting on @bratpiorka, @kilobyte, and @maciekpac)

___
*[Makefile.am, line 102 at r8](https://reviewable.io/reviews/memkind/memkind/786#-MzsifCLDSn8VrQl3GgQ:-MzsifCLDSn8VrQl3GgR:b-rt99dk) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/Makefile.am#L102)):*
> ```Makefile
>                   include/memkind/internal/mmap_tracing_queue.h \
>                   include/memkind/internal/slab_tracker.h \
>                   include/memkind/internal/timespec.h \
> ```

nit: looks like some standard header, which I assume it's not. Maybe we could rename to sth like timespec_ops.h? Same applies to the *.c file
___
*[src/mtt_allocator.c, line 80 at r8](https://reviewable.io/reviews/memkind/memkind/786#-Mzt3a_E1vodbNIymE5e:-Mzt3a_E1vodbNIymE5f:b-cz6cxb) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L80)):*
> ```C
>         bool flushing_rankings = false;
> 
>         pthread_mutex_lock(bg_thread_cond_mutex);
> ```

Can't we put pthread_cond_timedwait/clockwait here, right after this mutex lock ? The code would be shorter and we would avoid locking & unlocking a mutex.
___
*[src/mtt_allocator.c, line 92 at r8](https://reviewable.io/reviews/memkind/memkind/786#-MzsjP_4CqouzTUEPgVQ:-MzsjP_4CqouzTUEPgVR:bi4ejx2) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L92)):*
> ```C
>         // calculate time to wait
>         timespec_get_time(&start_time);
>         timespec_add(&start_time, &tv_period);
> ```

Here, start time no longer means start time, but the desired "end" time - name is misleading
___
*[src/mtt_allocator.c, line 109 at r8](https://reviewable.io/reviews/memkind/memkind/786#-MzsjcvQ6vRRon5l5WiC:-MzsjcvQ6vRRon5l5WiD:b-fp4h1a) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L109)):*
> ```C
> 
>         timespec_get_time(&end_time);
>         struct timespec wait_time = timespec_diff(&start_time, &end_time);
> ```

what if end_time > start time ? wouldn't we have integer overflow? If I'm correct, this scenario might really happen! (a lot of work done on bg thread)
___
*[src/mtt_allocator.c, line 110 at r8](https://reviewable.io/reviews/memkind/memkind/786#-Mzsjs2-9EWoQJCpQMDj:-Mzsjs2-9EWoQJCpQMDk:b-fp4h1a) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L110)):*
> ```C
>         timespec_get_time(&end_time);
>         struct timespec wait_time = timespec_diff(&start_time, &end_time);
>         if (timespec_not_zero(&wait_time)) {
> ```

what if end_time > start time ? wouldn't we have integer overflow? If I'm correct, this scenario might really happen! (a lot of work done on bg thread)
___
*[src/mtt_allocator.c, line 112 at r8](https://reviewable.io/reviews/memkind/memkind/786#-Mzt3NO629onc6wYi0Vv:-Mzt3NO79zRAngog8nc7:bo57cjs) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L112)):*
> ```C
>         if (timespec_not_zero(&wait_time)) {
>             // wait until next occurrence (according to frequency) or interrupt
>             pthread_mutex_lock(bg_thread_cond_mutex);
> ```

Can't it be put in the previous bg_thread_cond_mutex lock ? Please see another comment for reference.

**Review by bratpiorka — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 9 unresolved discussions (waiting on @kilobyte and @maciekpac)

**Review by maciekpac — COMMENTED** [review]
> Reviewed 1 of 9 files at r9.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 3 unresolved discussions (waiting on @bratpiorka, @kilobyte, and @maciekpac)

___
*[src/mtt_allocator.c, line 80 at r8](https://reviewable.io/reviews/memkind/memkind/786#-Mzt3a_E1vodbNIymE5e:-MzzfLn-5HHj_1ptpO5f:b-ledoug) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L80)):*
<details><summary><i>Previously, bratpiorka (Rafał Rudnicki) wrote…</i></summary><blockquote>

No, we have to wait at the end of this loop because the time spent on pebs_monitor and update_rankings should be included to waiting.
</blockquote></details>

I thought about including the time spent in pebs_monitor and update_rankings in the next iteration, but it's ok for me

**Review by bratpiorka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 3 unresolved discussions (waiting on @kilobyte and @maciekpac)

___
*[src/mtt_allocator.c, line 80 at r8](https://reviewable.io/reviews/memkind/memkind/786#-Mzt3a_E1vodbNIymE5e:-Mzzvgfy5Y05wwtDgXzN:b-896fix) ([raw file](https://github.com/memkind/memkind/blob/98a23e09177326e17b7f6bd63233dc6212e9cbfb/src/mtt_allocator.c#L80)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

I thought about including the time spent in pebs_monitor and update_rankings in the next iteration, but it's ok for me
</blockquote></details>

Done.

**Review by PatKamin — COMMENTED** [review]
> Reviewed 7 of 9 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 5 unresolved discussions (waiting on @bratpiorka, @kilobyte, and @maciekpac)

___
*[src/mtt_allocator.c, line 60 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzVssJX5AgrBRuL3OUy:-N-1mzJS1chaAHp3ag1P:bb5828c) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L60)):*
<details><summary><i>Previously, maciekpac wrote…</i></summary><blockquote>

nit: why do we declare a variable instead of directly using the macro ?
</blockquote></details>

nit: UP - maybe it's better to define `period_ms` instead of a frequency?
___
*[src/timespec_ops.c, line 26 at r10](https://reviewable.io/reviews/memkind/memkind/786#-N-1qT5o9D-JsNqSKaGH:-N-1qT5o9D-JsNqSKaGI:b-1djrtk) ([raw file](https://github.com/memkind/memkind/blob/98b29adfccab6a3e3fe09b0537a7376046de69d6/src/timespec_ops.c#L26)):*
> ```C
>     int ret = clock_gettime(clock_id, time);
>     if (ret != 0) {
>         log_fatal("ASSERT PEBS Monitor clock_gettime() FAILURE: %s",
> ```

nit: This function is not tied to the PEBS handling - it might be used anywhere in memkind. PEBS monitor shouldn't be mentioned in this error msg.
___
*[src/timespec_ops.c, line 35 at r10](https://reviewable.io/reviews/memkind/memkind/786#-N-1rOze0d8DjfGEV69a:-N-1rOzf4YPx2sZLn1xW:bgnr7hg) ([raw file](https://github.com/memkind/memkind/blob/98b29adfccab6a3e3fe09b0537a7376046de69d6/src/timespec_ops.c#L35)):*
> ```C
> {
>     out->tv_sec = (uint64_t)(millis / 1e3);
>     double ms_to_convert = millis - out->tv_sec * 1000ul;
> ```

nit: millis % 1e3

**Review by PatKamin — APPROVED** [review]
> Reviewed 1 of 9 files at r9.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 6 unresolved discussions (waiting on @bratpiorka, @kilobyte, and @maciekpac)

___
*[test/memkind_memtier_data_movement_test.cpp, line 763 at r10](https://reviewable.io/reviews/memkind/memkind/786#-N-1uhmJ13lYZ3KdJros:-N-1uhmJ13lYZ3KdJrot:b-mgrl1f) ([raw file](https://github.com/memkind/memkind/blob/98b29adfccab6a3e3fe09b0537a7376046de69d6/test/memkind_memtier_data_movement_test.cpp#L763)):*
> ```cpp
> 
>     for (size_t i = 0; i < size / sizeof(int); i++) {
>         tab[i] = i;
> ```

nit: I know that it's not a part of this commit, but assigning size_t to int happens here. Just to make sure, the value is irrelevant here?

**Review by bratpiorka — COMMENTED** [review]
> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 5 unresolved discussions (waiting on @kilobyte, @maciekpac, and @PatKamin)

___
*[src/mtt_allocator.c, line 60 at r4](https://reviewable.io/reviews/memkind/memkind/786#-MzVssJX5AgrBRuL3OUy:-N-25ZOv1h-TnLIGpBXR:b-wd62qo) ([raw file](https://github.com/memkind/memkind/blob/0b2c0113ec2d344bb5e3b15e5f474c12c05288c4/src/mtt_allocator.c#L60)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: UP - maybe it's better to define `period_ms` instead of a frequency?
</blockquote></details>

I think it is easier to use freq
___
*[src/timespec_ops.c, line 35 at r10](https://reviewable.io/reviews/memkind/memkind/786#-N-1rOze0d8DjfGEV69a:-N-269XZ-nOoO80KMnbN:b-x2obxv) ([raw file](https://github.com/memkind/memkind/blob/98b29adfccab6a3e3fe09b0537a7376046de69d6/src/timespec_ops.c#L35)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: millis % 1e3
</blockquote></details>

Please note, that millis is a double, so we can't mod it, and if we cast it to uint64_t, we lost the remainder part - so current implementation is more precise.
___
*[test/memkind_memtier_data_movement_test.cpp, line 763 at r10](https://reviewable.io/reviews/memkind/memkind/786#-N-1uhmJ13lYZ3KdJros:-N-26clZ7RwVvbugjJMB:bcvfnfl) ([raw file](https://github.com/memkind/memkind/blob/98b29adfccab6a3e3fe09b0537a7376046de69d6/test/memkind_memtier_data_movement_test.cpp#L763)):*
<details><summary><i>Previously, PatKamin (Patryk Kaminski) wrote…</i></summary><blockquote>

nit: I know that it's not a part of this commit, but assigning size_t to int happens here. Just to make sure, the value is irrelevant here?
</blockquote></details>

yes, this is just an initialization, the value is irrelevant

**Review by maciekpac — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 3 unresolved discussions (waiting on @kilobyte and @PatKamin)

**Review by PatKamin — APPROVED** [review]
> Reviewed 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/memkind/memkind/786)* status: all files reviewed, 1 unresolved discussion (waiting on @kilobyte)
